### PR TITLE
Add basic blocking tests

### PR DESCRIPTION
**Pull request by cataphract**

### CODE REVIEW

**Comment by cbeauchesne** [comment]
I've updated your branch, there were an issue fixed by #519

**Comment by cbeauchesne** [comment]
There is now an IP_BLOCKIG_SCENARIO, can I close this PR @cataphract ?

**Comment by cataphract** [comment]
> There is now an IP_BLOCKIG_SCENARIO, can I close this PR @cataphract ?

This one mainly tests that the response respects the Accept headers, so they're mostly orthogonal.

**Comment by cbeauchesne** [comment]
ok, so what can I do to help moving forward on this one ?

File: scenarios/appsec/waf/test_blocking.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        assert re.match("^application/json", r.headers["content-type"]) is not None
+        assert (
+            r.text == '{"errors": [{"title": "You\'ve been blocked", "detail": "Sorry, you cannot access '
+            'this page. Please contact the customer service team. Security provided by Datadog."}]}\n'
+        )
+
+    def test_blocking_appsec_blocked_tag(self):
+        """Tag ddappsec.blocked is set when blocking"""
+        r = self.weblog_get("/waf/", headers={"User-Agent": "Arachni/v1", "Accept": "*/*"})
+        assert r.status_code == 403
```
Comment (cbeauchesne) [inline scenarios/appsec/waf/test_blocking.py:141]:
```suggestion
        interfaces.library.add_assertion(r.status_code == 403)
```

And same for all others synchronous assertions

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
checking the failure

**Comment by cbeauchesne** [comment]
Seems to be good, just need to skiping open liberty

**Comment by cbeauchesne** [comment]
@cataphract does this PR still relevant ?

**Comment by cataphract** [comment]
> @cataphract does this PR still relevant ?

Sure. I've updated it. Can you approve and merge?

File: run.sh [hunk 2]
```diff
@@ -42,6 +42,11 @@ elif [ $SYSTEMTESTS_SCENARIO = "APPSEC_CORRUPTED_RULES" ]; then
 elif [ $SYSTEMTESTS_SCENARIO = "APPSEC_CUSTOM_RULES" ]; then
     WEBLOG_ENV+="DD_APPSEC_RULES=/appsec_custom_rules.json"
 
+elif [ $SYSTEMTESTS_SCENARIO = "APPSEC_BLOCKING" ]; then
+    export RUNNER_ARGS="tests/appsec/waf/test_blocking.py"
```
Comment (cbeauchesne) [inline run.sh:46]: Can you remove this, and add @scenario("APPSEC_BLOCKING") in all classes in `tests/appsec/waf/test_blocking.py` ?

File: run.sh [hunk 3]
```diff
@@ -42,6 +42,11 @@ elif [ $SYSTEMTESTS_SCENARIO = "APPSEC_CORRUPTED_RULES" ]; then
...
+    export SYSTEMTESTS_LOG_FOLDER=logs_blocking
```
Comment (cbeauchesne) [inline run.sh:47]: You can remove this line, it's automatic

File: tests/appsec/waf/test_blocking.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+)
+@coverage.basic
+@scenario("APPSEC_BLOCKING")
+class Test_Blocking:
+    """Blocking response is obtained when triggering a blocking rule"""
+
+    @bug(context.library < "java@0.115.0" and context.weblog_variant == "spring-boot-undertow", reason="npe")
+    def test_no_accept(self):
+        """Blocking without an accept header"""
+        r = weblog.get("/waf/", headers={"User-Agent": "Arachni/v1"})
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:131]: can you add a method `setup_no_accept` and send the request in it (save it as `self.r_not_accept`) ?

File: tests/appsec/waf/test_blocking.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            if "appsec.blocked" not in span["meta"]:
+                raise Exception("Can't find appsec.blocked in span's tags")
+
+            return True
+
+        interfaces.library.validate_spans(self.r_abt, validator=validate_appsec_blocked)
+
+    def test_accept_all(self):
+        """Blocking with Accept: */*"""
+        r = weblog.get("/waf/", headers={"User-Agent": "Arachni/v1", "Accept": "*/*"})
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:166]: Same comment, add a metohd `setup_accept_all` and send request inside it.

File: tests/appsec/waf/test_blocking.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        assert r.status_code == 403
+        assert re.match("^application/json", r.headers.get("content-type", "")) is not None
+
+    def test_accept_partial_json(self):
+        """Blocking with Accept: application/*"""
+        # */* should be ignored because there are more specific matches for text/html and application/json
+        r = weblog.get(
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:173]: Idem

File: tests/appsec/waf/test_blocking.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+            "/waf/", headers={"User-Agent": "Arachni/v1", "Accept": "text/*;q=0.7, application/*;q=0.8, */*;q=0.9"}
+        )
+        assert r.status_code == 403
+        assert re.match("^application/json", r.headers.get("content-type", "")) is not None
+
+    def test_accept_partial_html(self):
+        """Blocking with Accept: text/*"""
+        r = weblog.get(
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:181]: idem

File: tests/appsec/waf/test_blocking.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            "/waf/", headers={"User-Agent": "Arachni/v1", "Accept": "text/*;q=0.8, application/*;q=0.7, */*;q=0.9"}
+        )
+        assert r.status_code == 403
+        assert r.text == HTML_DATA
+
+    def test_accept_full_json(self):
+        """Blocking with Accept: application/json"""
+        r = weblog.get(
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:189]: idem

File: tests/appsec/waf/test_blocking.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+                "User-Agent": "Arachni/v1",
+                "Accept": "text/*;q=0.8, application/*;q=0.7, application/json;q=0.85, */*;q=0.9",
+            },
+        )
+        assert r.status_code == 403
+        assert re.match("^application/json", r.headers.get("content-type", "")) is not None
+
+    def test_accept_full_html(self):
+        """Blocking with Accept: text/html"""
+        r = weblog.get(
```
Comment (cbeauchesne) [inline tests/appsec/waf/test_blocking.py:201]: idem

File: tests/test_standard_tags.py [hunk 10]
```diff
@@ -38,7 +38,10 @@ def test_method_trace(self):
 @released(dotnet="2.13.0", golang="1.40.0", java="0.107.1", nodejs="3.0.0")
 @released(php="0.76.0", python="1.6.0rc1.dev", ruby="?")
 @rfc("https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2490990623/QueryString+-+Sensitive+Data+Obfuscation")
-@bug(weblog_variant="spring-boot-undertow", reason="APMJAVA-877")
+@bug(
```
Comment (cbeauchesne) [inline tests/test_standard_tags.py:41]: is it relevant in this PR? if I'm not wrong, it's already fixed skipped no ?

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
I've converted the PR as a draft (we ares spammed on slack by PR to be reviwed. Please set it back to ready once comment are fixed :)

File: tests/test_standard_tags.py [hunk 11]
```diff
@@ -38,7 +38,6 @@ def test_method_trace(self):
 @released(dotnet="2.13.0", golang="1.40.0", java="0.107.1", nodejs="3.0.0")
 @released(php="0.76.0", python="1.6.0rc1.dev", ruby="?")
 @rfc("https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2490990623/QueryString+-+Sensitive+Data+Obfuscation")
-@bug(weblog_variant="spring-boot-undertow", reason="APMJAVA-877")
```
Comment (cbeauchesne) [inline tests/test_standard_tags.py:41]: MMh, now it's removed? I may have added a dummy comment, or is it some merge issue ?

File: utils/scripts/generate-github-action-file.py [hunk 12]
```diff
@@ -216,6 +217,8 @@ def add_main_job(i, workflow, needs, scenarios):
 
         if scenario == "TRACE_PROPAGATION_STYLE_W3C":  # TODO: fix weblog to allow this value for old tracer
             step["if"] = "${{ matrix.variant.library != 'python' }}"  # TODO
+        elif scenario == "APPSEC_BLOCKING":
+            step["if"] = "${{ matrix.library == 'java' }}"
```
Comment (cbeauchesne) [inline utils/scripts/generate-github-action-file.py:221]: Why? if the test if flagged as fail, it's ok to run it.

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: utils/scripts/generate-github-action-file.py (hunk 12 above)
Comment (cataphract) [reply utils/scripts/generate-github-action-file.py:221]: We can run it, but we're gonna waste CPU because it won't pass outside java

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**